library/include/BitArray.h

```diff
@@ -526,27 +407,125 @@ namespace DFHack
 
         iterator begin()
         {
-            return iterator(static_cast<L *>(this), static_cast<L *>(this));
+            return iterator(static_cast<L *>(this), static_cast<L *>(this)->next);
         }
         const_iterator begin() const
         {
-            return const_iterator(static_cast<L *>(this), static_cast<L *>(this));
+            return const_iterator(static_cast<L *>(this), static_cast<L *>(this)->next);
         }
         const_iterator cbegin() const
         {
             return begin();
         }
         iterator end()
         {
-            return iterator(static_cast<L *>(this), nullptr, false);
+            return iterator(static_cast<L *>(this), nullptr);
         }
         const_iterator end() const
         {
-            return const_iterator(static_cast<L *>(this), nullptr, false);
+            return const_iterator(static_cast<L *>(this), nullptr);
         }
         const_iterator cend() const
         {
             return end();
         }
+
+        iterator erase(const_iterator pos)
+        {
+            auto root = static_cast<L *>(this);
+            CHECK_INVALID_ARGUMENT(pos.iter.root == root);
+            CHECK_NULL_POINTER(pos.iter.cur);
+
+            auto link = pos.iter.cur;
+            auto next = link->next;
+            if (link->next && link->prev)
+            {
+                link->next->prev = link->prev;
+                link->prev->next = link->next;
+            }
+            else if (link->next)
+            {
+                link->next->prev = nullptr;
+            }
+            else if (link->prev)
+            {
+                link->prev->next = nullptr;
+            }
+
+            if (root->next == link)
+            {
+                root->next = next;
+            }
+
+            delete link;
+
+            return iterator(root, next);
+        }
+        iterator insert(const_iterator pos, I * const & item)
+        {
+            auto root = static_cast<L *>(this);
+            CHECK_INVALID_ARGUMENT(pos.iter.root == root);
+
+            auto link = pos.iter.cur;
+            if (!link || !link->prev)
+            {
+                if (!link && root->next)
+                {
+                    pos--;
+                    return insert_after(pos, item);
+                }
+
+                CHECK_INVALID_ARGUMENT(root->next == link);
+                push_front(item);
+                return begin();
+            }
+
+            auto newlink = new L();
+            newlink->prev = link->prev;
+            newlink->next = link;
+            link->prev = newlink;
+            if (newlink->prev)
+            {
+                newlink->prev->next = newlink;
+            }
+            else if (link == root->next)
+            {
+                root->next = newlink;
+            }
+            newlink->item = item;
+            return iterator(root, newlink);
```

Comment:
I wonder if insert could work even if link->prev == nullptr (inserting to the begin). But not if iterator is end().
<pre>
newlink->prev = link->prev;
newlink->next = link;
if (link->prev) link->prev->next = newlink;
else root->next = newlink;
link->prev = newlink;
</pre>